### ISSUE TITLE
Implement cons

### ISSUE THREAD

**Issue by Wilfred** [Wilfred/remacs#73]
This will require some allocation work.

**Comment by crlf0710** [comment]
How is it going? Is there anything i can help?

**Comment by Wilfred** [comment]
Having too much fun on other Remacs things and helping land PRs :)

Anyway, I've mostly got this code sorted on my cons_function branch. However, I need to define a cons block type:

```rust
#[repr(C)]
struct ConsBlock {
    conses: [LispCons; CONS_BLOCK_SIZE as usize],
    gcmarkbits: [bits_word; (1 + CONS_BLOCK_SIZE / BITS_PER_BITS_WORD) as usize],
    next: *mut ConsBlock,
}
```

The problem is defining `CONS_BLOCK_SIZE`. This is defined in alloc.c:

```c
#define CONS_BLOCK_SIZE						\
  (((BLOCK_BYTES - sizeof (struct cons_block *)			\
     /* The compiler might add padding at the end.  */		\
     - (sizeof (struct Lisp_Cons) - sizeof (bits_word))) * CHAR_BIT)	\
   / (sizeof (struct Lisp_Cons) * CHAR_BIT + 1))
``

which in turn depends on `BLOCK_BYTES`:

```c
#define BLOCK_BYTES \
  (BLOCK_ALIGN - sizeof (struct ablocks *) - BLOCK_PADDING)
```

and I reach a point where I need to implement significant parts of alloc.c!

I've been distracted by other things since. I think the best thing to do right now would be to hard-code the value of `CONS_BLOCK_SIZE` to whatever the C macro evaluates to.

I'm open to suggestions though :)

**Comment by crlf0710** [comment]
Is it good to just define a variable (constant) in the C side and pick it up from the Rust side?

**Comment by Wilfred** [comment]
That's a good idea, but sadly I don't think it works. Rust doesn't let you define constants in extern blocks, so this isn't legal:

```rust
extern "C" {
    const CONS_BLOCK_SIZE: libc::c_int;
}
```

I could use a `static`:

```rust
extern "C" {
    static CONS_BLOCK_SIZE: libc::c_int;
}
```

however this is no longer a constant expression so I can't use it in type declarations.

**Comment by ljos** [comment]
I have been looking into the alloc and garbage collection code in alloc.c and I don't know if you need to implement most of this in the exact same way as they do there. 

The reason for the `cons_block` structure is to make it faster to allocate new `cons`. It seems to me to also be private. If you look at [lisp.h#L3490-L3669](https://github.com/Wilfred/remacs/blob/666a5c553338dd1ee2114196b5ec3dc7c9f0cb7c/src/lisp.h#L3490-L3669) and [alloc.c#L7408-L7526](https://github.com/Wilfred/remacs/blob/666a5c553338dd1ee2114196b5ec3dc7c9f0cb7c/src/alloc.c#L7408-L7526); these are the public functions from alloc.c we need to preserve to be interoperable with Emacs. We don't need to preserve the internal structures of the GC. They are not used in other parts of the C code and is not exposed in any header file.

It also looks to me that a lot of the code concerning calculating the alignment of structs, padding, etc, is better done by the system malloc. A lot of that code is because Emacs supports multiple different mallocs and, since it is very old, doesn't trust malloc to align data properly. I don't think we need to do that. We can let the malloc in rust do these things for us. 

We have the problem that now the macOS version uses a (sort of) custom malloc, but that won't be the case when we remove unexec.

TL;DR:
You don't need to concern yourself (too much) with how Emacs implements cons_block, CONS_BLOCK_SIZE, etc, as they are internal to the garbage collector.
